# Online Fine-Tuning of a BERT Model for Continual Classification

## Introduction
In dynamic environments where data is continuously generated (e.g., daily quotes), an online or continual learning approach enables a pre-trained BERT model to be incrementally fine-tuned. This method allows the model to adapt to evolving language patterns and data distributions while retaining previously learned knowledge.

## Methodology
1. **Initial Fine-Tuning:**
   - Start with a pre-trained BERT model.
   - Fine-tune it on the initial dataset for the target classification task (binary or multiclass).

2. **Continual Updates:**
   - As new data arrives (e.g., daily quotes), periodically fine-tune the existing model using the new data.
   - Fine-tuning is done in batches (e.g., daily or weekly) rather than on a per-sample basis.
   - Optionally, mix new data with a subset of historical data to preserve prior knowledge.

3. **Pipeline Setup:**
   - Establish a regular schedule for model updates and performance evaluation.
   - Continue training from the current state of the model using transfer learning principles.

## Considerations
- **Catastrophic Forgetting:**
  - *Problem:* Fine-tuning on new data can lead the model to forget previously learned information.
  - *Mitigation Strategies:*
    - **Rehearsal:** Include a subset of historical data in each update.
    - **Regularization:** Use methods like Elastic Weight Consolidation (EWC) to prevent drastic changes in important parameters.
      - Reference: [Kirkpatrick et al., 2017](https://www.pnas.org/doi/10.1073/pnas.1611835114)
    - **Memory-Based Approaches:** Maintain a small buffer of past examples to be revisited during fine-tuning.

- **Hyperparameter Tuning:**
  - Adjust learning rates, batch sizes, and the number of epochs based on the new data.
  - Monitor performance on both recent and historical validation sets to ensure balanced learning.

- **Data Distribution Shifts:**
  - Continuously monitor for changes in data characteristics.
  - Adapt the update strategy if significant shifts in the distribution are detected.

- **Computational Resources:**
  - Online fine-tuning can be resource-intensive.
  - Optimize update frequency and batch sizes to maintain a balance between performance improvements and resource usage.

## Conclusion
Online fine-tuning of a BERT model is an effective strategy for maintaining up-to-date classifiers in environments with continuous data influx. By scheduling periodic updates and employing techniques to mitigate catastrophic forgetting, you can ensure that the model remains robust and performs well on both new and previously seen data.

## References
- Kirkpatrick, J., Pascanu, R., Rabinowitz, N., Veness, J., Desjardins, G., Rusu, A. A., ... & Hadsell, R. (2017). **Overcoming catastrophic forgetting in neural networks**. *Proceedings of the National Academy of Sciences, 114*(13), 3521-3526. [Link](https://www.pnas.org/doi/10.1073/pnas.1611835114)
- Parisi, G. I., Kemker, R., Part, J. L., Kanan, C., & Wermter, S. (2019). **Continual lifelong learning with neural networks: A review**. *Neural Networks, 113*, 54-71. [Link](https://doi.org/10.1016/j.neunet.2019.01.012)
- Hugging Face Transformers Documentation. [Link](https://huggingface.co/transformers/)


In [2]:
import pandas as pd
df_train = pd.read_csv('df_train_sbert.csv')
df_test = pd.read_csv('df_test_sbert.csv')

In [7]:
pip install --upgrade pip


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 7.4 MB/s eta 0:00:00a 0:00:01
  Attempting uninstall: pip
    Found existing installation: pip 25.0
    Uninstalling pip-25.0:
      Successfully uninstalled pip-25.0
Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.7/9.7 MB 6.1 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 9.7 MB/s eta 0:00:00ta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip install tf-keras


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 5.3 MB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [7]:
pip install datasets

Note: you may need to restart the kernel to use updated packages.


In [11]:
pip install transformers[torch]

zsh:1: no matches found: transformers[torch]


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Note: you may need to restart the kernel to use updated packages.


In [14]:
pip install 'accelerate>=0.26.0'

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Note: you may need to restart the kernel to use updated packages.


In [3]:
from transformers import BertTokenizerFast, BertForSequenceClassification, Trainer, TrainingArguments
from sklearn.model_selection import train_test_split
import numpy as np
from datasets import Dataset

# Assume df_train and df_test already exist and have been preprocessed with:
# - 'processed_text': the preprocessed text from your pipeline.
# - 'label': binary classification label (0 for Out-of-Topic, 1 for In-Topic)


# Convert the Pandas DataFrames to Hugging Face Datasets.
train_dataset = Dataset.from_pandas(df_train)
test_dataset  = Dataset.from_pandas(df_test)

# Load the pre-trained BERT tokenizer.
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')

# Define tokenization function using the 'processed_text' column.
def tokenize_function(examples):
    texts = [str(text) for text in examples['processed_text']]
    return tokenizer(texts, padding='max_length', truncation=True, max_length=128)


# Tokenize the datasets.
train_dataset = train_dataset.map(tokenize_function, batched=True)
test_dataset  = test_dataset.map(tokenize_function, batched=True)

# Set format for PyTorch.
train_dataset.set_format(type='torch', columns=['input_ids', 'attention_mask', 'label'])
test_dataset.set_format(type='torch', columns=['input_ids', 'attention_mask', 'label'])

# Initialize the BERT model for binary classification (num_labels=2).
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2)

# Define training arguments.
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=32,
    warmup_steps=500,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    logging_dir='./logs',
    logging_steps=10,
    load_best_model_at_end=True,
    metric_for_best_model="accuracy",
)

# Define a simple accuracy metric.
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    accuracy = np.mean(predictions == labels)
    return {"accuracy": accuracy}

# Initialize the Trainer.
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    compute_metrics=compute_metrics,
)

# Fine-tune the binary classification model.
trainer.train()

# Evaluate the model.
eval_results = trainer.evaluate()
print("Evaluation results:", eval_results)


Map:   0%|          | 0/88911 [00:00<?, ? examples/s]

Map:   0%|          | 0/22228 [00:00<?, ? examples/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/opt/anaconda3/lib/python3.12/site-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Evaluation results: {'eval_loss': 0.152065247297287, 'eval_accuracy': 0.9484883930178154, 'eval_runtime': 372.3748, 'eval_samples_per_second': 59.693, 'eval_steps_per_second': 1.866, 'epoch': 3.0}


In [4]:
predictions_output = trainer.predict(test_dataset)
print("Predictions:", predictions_output.predictions)
print("Labels:", predictions_output.label_ids)


Predictions: [[ 4.7244315 -4.2403507]
 [ 4.7080913 -4.238422 ]
 [ 4.4483476 -4.126842 ]
 ...
 [-1.6329443  1.8176602]
 [ 4.737137  -4.2433586]
 [ 4.6952157 -4.232057 ]]
Labels: [0 0 0 ... 1 0 0]


In [5]:
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix

# Assume predictions_output is the result from trainer.predict(test_dataset)
# For models that output logits, use argmax to convert to predicted labels.
predictions_output = trainer.predict(test_dataset)
predicted_labels = np.argmax(predictions_output.predictions, axis=1)
true_labels = predictions_output.label_ids

print("Classification Report:")
print(classification_report(true_labels, predicted_labels))
print("Confusion Matrix:")
print(confusion_matrix(true_labels, predicted_labels))


Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.96      0.97     18301
           1       0.82      0.90      0.86      3927

    accuracy                           0.95     22228
   macro avg       0.90      0.93      0.91     22228
weighted avg       0.95      0.95      0.95     22228

Confusion Matrix:
[[17548   753]
 [  392  3535]]


In [15]:
import numpy as np
import pandas as pd

# ---- Step 1: Extract Predicted Positives from Binary Classifier ----
X_train_bin = df_train["sbert_embedding"]
y_train_bin = df_train["label"]
# - binary_predict is defined from the upstream task

# Get binary predictions on the training set
# from the upstream task
binary_preds_output = trainer.predict(train_dataset)
binary_preds_train = np.argmax(binary_preds_output.predictions, axis=1) # array of 0s and 1s

# Find indices where the binary classifier predicts positive (in-topic)
positive_indices = np.where(binary_preds_train == 1)[0]

print("Number of samples predicted as in-topic:", len(positive_indices))

# ---- Step 2: Build a New Training Set for the Downstream Classifier ----
# We'll extract rows from df_train corresponding to predicted positives.
# Then, for each extracted sample:
#   - If the true binary label is 1 (i.e., it is a true positive), keep its original 'topic_id'
#   - If the true binary label is 0 (i.e., a false positive), set its 'topic_id' to "NP"

df_downstream = df_train.iloc[positive_indices].copy()

# Create a new column for the downstream topic label:
df_downstream['downstream_topic'] = df_downstream.apply(
    lambda row: row['topic_id'] if row['label'] == 1 else 0.0, axis=1
)

# Now, df_downstream contains only the samples predicted as in-topic.
# Their 'downstream_topic' column holds the original topic for true positives,
# and "NP" for false positives.

print("Downstream training set shape:", df_downstream.shape)
print("Value counts for downstream topics:")
print(df_downstream['downstream_topic'].value_counts())

# ---- Step 3: (Optional) Prepare Data for Downstream BERT Fine-Tuning ----
# For instance, if you want to fine-tune a BERT classifier on this subset:
# Make sure your downstream training set contains:
# - 'processed_text': the input text.
# - 'downstream_topic': the new multiclass labels (including "NP").

# You might need to remap 'downstream_topic' to contiguous integers, for example:
unique_topics = np.sort(df_downstream['downstream_topic'].unique())
topic_mapping = {topic: idx for idx, topic in enumerate(unique_topics)}
df_downstream['mapped_topic'] = df_downstream['downstream_topic'].map(topic_mapping)

print("Unique downstream topics mapping:", topic_mapping)

# At this point, you can use df_downstream to train your downstream classifier.
# For example, you could convert it to a Hugging Face Dataset and fine-tune a BERT model:
from datasets import Dataset
downstream_dataset = Dataset.from_pandas(df_downstream)

# Then tokenize using your usual tokenization function,
# and fine-tune a BERT model (or DistilBERT, etc.) on 'processed_text' with labels 'mapped_topic'.

# (Fine-tuning code for BERT on downstream_dataset would go here.)



Number of samples predicted as in-topic: 16767
Downstream training set shape: (16767, 14)
Value counts for downstream topics:
downstream_topic
602.0    2617
543.0    2327
546.0    2261
0.0      2181
544.0    2132
550.0    2062
547.0    1490
600.0     881
554.0     359
552.0     230
556.0     227
Name: count, dtype: int64
Unique downstream topics mapping: {0.0: 0, 543.0: 1, 544.0: 2, 546.0: 3, 547.0: 4, 550.0: 5, 552.0: 6, 554.0: 7, 556.0: 8, 600.0: 9, 602.0: 10}


In [17]:
print(downstream_dataset)

Dataset({
    features: ['Unnamed: 0', 'country_id', 'country_name', 'product_id', 'product_category', 'review_id', 'review_text', 'quote_text', 'quote_id', 'topic_id', 'label', 'processed_text', 'sbert_embedding', 'downstream_topic', 'mapped_topic', '__index_level_0__'],
    num_rows: 16767
})


In [18]:
import numpy as np
import pandas as pd
from transformers import BertTokenizerFast, BertForSequenceClassification, Trainer, TrainingArguments
from datasets import Dataset
from sklearn.metrics import classification_report, confusion_matrix

# -------------
# ASSUMPTIONS:
# - df_train and df_test are for the binary task and include:
#      'sbert_embedding': SBERT embeddings (lists/arrays)
#      'label': binary labels (0 for out-of-topic, 1 for in-topic)
# - df_downstream is for the downstream (multiclass) task and includes:
#      'sbert_embedding': SBERT embeddings (lists/arrays)
#      'topic_id': the original topic label for in-topic samples; NaN for out-of-topic.
#
# - topic_mapping is a dictionary mapping the original topic IDs (plus the out-of-topic placeholder)
#   to contiguous integers.
# -------------

# --- STEP 1: Extract downstream test set from df_test based on binary predictions ---

# Get binary predictions on df_test using your binary model Trainer (assumed already trained).
# This returns an object; we extract predictions and then take argmax to get 0/1.
binary_predictions_output = trainer.predict(test_dataset)  # 'trainer' is your binary model Trainer
binary_preds_test = np.argmax(binary_predictions_output.predictions, axis=1)

# Find indices where the binary classifier predicts in-topic (1)
positive_indices_test = np.where(binary_preds_test == 1)[0]
print("Number of test samples predicted as in-topic:", len(positive_indices_test))

# Create downstream test DataFrame from df_test (for multiclass stage).
df_downstream_test = df_test.iloc[positive_indices_test].copy()

# Create a new column 'downstream_topic':
# If the true binary label is 1, use the true 'topic_id'; otherwise, mark as "NP".
df_downstream_test['downstream_topic'] = df_downstream_test.apply(
    lambda row: row['topic_id'] if row['label'] == 1 else 0.0, axis=1
)

# Map the downstream_topic to contiguous integers using topic_mapping.
# (Ensure that topic_mapping is defined; for example, you might have built it from df_downstream.)
df_downstream_test['mapped_topic'] = df_downstream_test['downstream_topic'].map(topic_mapping)

# Create a Hugging Face Dataset for the downstream test set.
downstream_dataset_test = Dataset.from_pandas(df_downstream_test)

# --- STEP 2: Prepare the Downstream Training Data ---
# Assume df_downstream is already prepared similarly (it includes 'processed_text', 'downstream_topic', and 'mapped_topic')
downstream_dataset = Dataset.from_pandas(df_downstream)

# --- STEP 3: Tokenization ---
# Load the BERT tokenizer.
tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")

def tokenize_function(examples):
    # Ensure text is a string.
    texts = [str(text) for text in examples["processed_text"]]
    return tokenizer(texts, padding="max_length", truncation=True, max_length=128)

# Tokenize training and test downstream datasets.
downstream_train = downstream_dataset.map(tokenize_function, batched=True)
downstream_test  = downstream_dataset_test.map(tokenize_function, batched=True)

# Set the format for PyTorch; include the label column ('mapped_topic').
downstream_train.set_format(type="torch", columns=["input_ids", "attention_mask", "mapped_topic"])
downstream_test.set_format(type="torch", columns=["input_ids", "attention_mask", "mapped_topic"])

# Determine the number of downstream classes.
num_downstream_classes = len(np.unique(df_downstream["mapped_topic"].values))
print("Number of downstream classes:", num_downstream_classes)

# Rename the label column to "labels" for Trainer compatibility.
downstream_train = downstream_train.rename_column("mapped_topic", "labels")
downstream_test  = downstream_test.rename_column("mapped_topic", "labels")

# --- STEP 4: Downstream BERT Fine-Tuning ---
model_downstream = BertForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=num_downstream_classes)

training_args = TrainingArguments(
    output_dir="./results_downstream",
    num_train_epochs=3,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=32,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    logging_dir="./logs_downstream",
    logging_steps=10,
    load_best_model_at_end=True,
    metric_for_best_model="accuracy",
)

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    accuracy = np.mean(predictions == labels)
    return {"accuracy": accuracy}

trainer_downstream = Trainer(
    model=model_downstream,
    args=training_args,
    train_dataset=downstream_train,
    eval_dataset=downstream_test,
    compute_metrics=compute_metrics,
)

# Fine-tune the downstream BERT model.
trainer_downstream.train()

# Evaluate the downstream model.
eval_results_downstream = trainer_downstream.evaluate()
print("Downstream Evaluation results:", eval_results_downstream)

predictions_output = trainer_downstream.predict(downstream_test)
predicted_labels = np.argmax(predictions_output.predictions, axis=1)
true_labels = predictions_output.label_ids

print("Downstream Classification Report:")
print(classification_report(true_labels, predicted_labels))
print("Downstream Confusion Matrix:")
print(confusion_matrix(true_labels, predicted_labels))


Number of test samples predicted as in-topic: 4288


Map:   0%|          | 0/16767 [00:00<?, ? examples/s]

Map:   0%|          | 0/4288 [00:00<?, ? examples/s]

Number of downstream classes: 11


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/opt/anaconda3/lib/python3.12/site-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Downstream Evaluation results: {'eval_loss': 0.6768686175346375, 'eval_accuracy': 0.7590951492537313, 'eval_runtime': 68.9176, 'eval_samples_per_second': 62.219, 'eval_steps_per_second': 1.944, 'epoch': 3.0}
Downstream Classification Report:
              precision    recall  f1-score   support

           0       0.53      0.16      0.25       753
           1       0.77      0.92      0.84       561
           2       0.79      0.51      0.62       499
           3       0.78      0.92      0.84       529
           4       0.87      0.97      0.92       404
           5       0.68      0.96      0.80       502
           6       0.77      0.91      0.83        58
           7       0.79      0.87      0.83        78
           8       0.79      0.95      0.86        40
           9       0.85      0.99      0.92       211
          10       0.76      0.98      0.86       653

    accuracy                           0.76      4288
   macro avg       0.76      0.83      0.78      4288


In [21]:
# Save the binary classifier model
trainer.model.save_pretrained("./binary_bert_model")

# Save the tokenizer, same for downstream and upstream
tokenizer.save_pretrained("./binary_bert_model")

# Save the downstream classifier model
trainer_downstream.model.save_pretrained("./downstream_bert_model")

# Save the tokenizer (it's the same tokenizer, but good practice to save it again)
tokenizer.save_pretrained("./downstream_bert_model")

('./downstream_bert_model/tokenizer_config.json',
 './downstream_bert_model/special_tokens_map.json',
 './downstream_bert_model/vocab.txt',
 './downstream_bert_model/added_tokens.json',
 './downstream_bert_model/tokenizer.json')

In [ ]:
# to reload
from transformers import BertForSequenceClassification, BertTokenizerFast

# Load the downstream classifier model
downstream_model = BertForSequenceClassification.from_pretrained("./downstream_bert_model")

# Load the tokenizer
tokenizer = BertTokenizerFast.from_pretrained("./downstream_bert_model")


# Load the binary classifier model
binary_model = BertForSequenceClassification.from_pretrained("./binary_bert_model")




# **Analysis of BERT-on-BERT Downstream Performance**
Your **BERT-based binary model** has already demonstrated strong performance with **95% accuracy**, and now we are evaluating its **downstream BERT multiclass classifier**.

## **1. Key Overall Metrics**
- **Loss**: `0.6769` (acceptable, but could be improved)
- **Accuracy**: **`76%`** (solid, similar to previous methods but with better F1-macro)
- **Macro F1 Score**: **`0.78`** (better than the XGB/SVM pipeline, which was around `0.62–0.64`)
- **Weighted F1 Score**: **`0.72`** (shows overall balance in classification)

---

## **2. Breakdown of Precision, Recall, and F1**
| Topic ID | Precision | Recall | F1-score | Support | Notes |
|----------|-----------|--------|----------|---------|-------|
| **0**  | **0.53** | **0.16** | **0.25** | 753 | Many false positives from binary stage end up here |
| **1**  | **0.77** | **0.92** | **0.84** | 561 | Very high recall, good precision |
| **2**  | **0.79** | **0.51** | **0.62** | 499 | Precision is good, but recall is low |
| **3**  | **0.78** | **0.92** | **0.84** | 529 | Strong recall and F1 |
| **4**  | **0.87** | **0.97** | **0.92** | 404 | Very high precision and recall |
| **5**  | **0.68** | **0.96** | **0.80** | 502 | Recall is much higher than precision |
| **6**  | **0.77** | **0.91** | **0.83** | 58  | Very high recall |
| **7**  | **0.79** | **0.87** | **0.83** | 78  | Consistently strong performance |
| **8**  | **0.79** | **0.95** | **0.86** | 40  | Excellent recall |
| **9**  | **0.85** | **0.99** | **0.92** | 211 | Almost perfect recall |
| **10** | **0.76** | **0.98** | **0.86** | 653 | Very strong overall |

---

## **3. Observations from Confusion Matrix**
1. **Class 0 is still problematic** (False Positives from Binary Stage)
   - **Recall = 16%**, meaning **many misclassified out-of-topic quotes are still incorrectly labeled as other topics.**
   - Many examples from **topics 1, 2, 3, 5, and 10** are mistakenly categorized as **class 0**.
   - A **better binary threshold** or **a filtering mechanism for low-confidence predictions** may help.

2. **Class Imbalance is Well Handled**
   - **Rare classes (e.g., 6, 7, 8, 9, 10) have good recall (~90%)**.
   - Unlike SVM, which struggled with small classes, **BERT generalizes much better**.

3. **Most Topics Have >90% Recall**
   - Topics **1, 3, 4, 5, 6, 7, 8, 9, and 10** are very well classified, with recall approaching **97–99%** in some cases.
   - Precision is slightly lower in topics like **5 and 6**, meaning there are some false positives.

4. **Topic 2 is an Outlier**
   - **Precision: 79%** (good)
   - **Recall: 51%** (low)
   - It suggests **many real topic 2 quotes were misclassified**. Some were misclassified as **topic 5 or topic 10**.
   - Possible fix: **more training data for topic 2** or **better class weighting in loss function**.

---

## **4. How This Compares to XGB+SVM**
| Metric | **BERT-on-BERT** | **XGB+SVM** |
|--------|----------------|-------------|
| **Binary Accuracy** | **95%** | 86% |
| **Multiclass Accuracy** | **76%** | 76% |
| **Macro F1 (Multiclass)** | **0.78** | 0.62–0.64 |
| **Weighted F1** | **0.72** | 0.64 |
| **Recall (Multiclass)** | **0.83** | ~0.62 |
| **Worst-Class Recall (Class 0 / False Positives from Binary)** | **16%** | ~10% |

- **BERT-on-BERT maintains accuracy while massively improving F1.**
- **Recall is much higher** in most categories, making **BERT generalizes much better than SVM.**
- **Binary model improvements could help further**.

---

## **5. Next Steps to Improve BERT-on-BERT**
1. **Reduce False Positives from the Binary Model**
   - Class **0 recall is too low (16%)**, meaning some **out-of-topic quotes are misclassified in binary** and not corrected later.
   - Try **adjusting the binary threshold** or **using confidence scores**.
   - Another option: **train a secondary "uncertainty filter" to reduce noisy predictions**.

2. **Balance Class Distribution**
   - Class **2 (low recall)** might benefit from **more training data or class weighting**.
   - If certain classes are **overrepresented in training**, the model may be biased.

3. **Use a Better Loss Function**
   - Try **Focal Loss** to **handle difficult classes (e.g., topic 2) more effectively**.
   - This helps address class imbalance without manually oversampling.

4. **Consider RoBERTa or DeBERTa for Even Higher Performance**
   - If you want **even better accuracy**, **RoBERTa or DeBERTa** might help.
   - They **outperform BERT in many classification tasks**.

---

## **6. Final Verdict**
**✅ BERT-on-BERT is the best approach you've tested.**  
- It **outperforms XGB+SVM in every aspect**.
- It has **better recall across most classes**, making it **more reliable for topic classification**.
- Some minor issues remain, but **adjusting binary filtering & balancing training data** could push performance even higher.

### **BERT-on-BERT WINS 🎉**
